### PR TITLE
Fix firewall (the return)

### PR DESCRIPTION
**Pull request by outscale-jju**

### CODE REVIEW

File: api/server/graph.cc [hunk 1]
```diff
@@ -608,9 +608,19 @@ std::string Graph::fw_build_rule(const app::Rule &rule) {
 std::string Graph::fw_build_sg(const app::Sg &sg) {
     std::string r;
     for (auto it = sg.rules.begin(); it != sg.rules.end();) {
-        r += "(" + fw_build_rule(it->second) + ")";
+        auto fw_rule = fw_build_rule(it->second);
```
Comment (outscale-mgo) [inline api/server/graph.cc]: do we really need to use auto here ?
I mean, I have no idea what's the type of fw_rule.

File: api/server/graph.cc [hunk 2]
```diff
@@ -639,9 +649,19 @@ void Graph::fw_update(const app::Nic &nic) {
             it++;
             continue;
         }
-        in_rules += "( " + fw_build_sg(sit->second) + " )";
+        auto sg_rules = fw_build_sg(sit->second);
```
Comment (outscale-mgo) [inline api/server/graph.cc]: same here, please don't use auto, when the type is not explicit, I mean in a for with an iterator, the type is pretty explicit, when follow by a new, the type is pretty explicit too, but here, I need to fw_build_sg if I want to understand what sg_rules is, and that make the code harder to read.

File: api/server/graph.cc [hunk 3]
```diff
@@ -608,9 +608,19 @@ std::string Graph::fw_build_rule(const app::Rule &rule) {
...
+            it++;
+            continue;
+        }
+        r += "(" + fw_rule + ")";
         if (++it != sg.rules.end())
-            r += " || ";
+            r += "||";
+    }
+    // Special case when last rule is empty
+    if (r.back() == '|') {
```
Comment (outscale-mgo) [inline api/server/graph.cc]: how could this happen ? if a rule is empty you skip the rule with a continue ...

File: api/server/graph.cc [hunk 4]
```diff
@@ -632,9 +649,19 @@ void Graph::fw_update(const app::Nic &nic) {
...
+            it++;
+            continue;
+        }
+        in_rules += "(" + fw_build_sg(sit->second) + ")";
         if (++it != nic.security_groups.end())
-            in_rules += " || ";
+            in_rules += "||";
+    }
+    // Special case when last rule is empty
+    if (in_rules.back() == '|') {
```
Comment (outscale-mgo) [inline api/server/graph.cc]: same here, can this condition really happen ?

**Comment by outscale-mgo** [comment]
I've push the patch 2 and 3 but not "api/server: fix non-compilable rule"

**Comment by outscale-jju** [comment]
You did not merged `api/server: fix non-compilable rule`

**Comment by outscale-jju** [comment]
merged

**Pull request closed without merging**